File: JanperCab/5-JanperCab/ClientApp/src/app/_models/customer/CustomerDto.ts
import { CustomerCategoryCBDDto } from './../customer-category/CustomerCategoryCBDDto';
import { CustomerCategoryAccountDto } from './../customer-category/CustomerCategoryAccountDto';
import { Type } from 'class-transformer';
import { CustomerCategoryDto } from '../customer-category/CustomerCategoryDto';
import { CustomerType } from './../../_enums/CustomerType';
import { CabinetMakerDto } from './CabinetMakerDto';
import { DistributorDto } from './DistributorDto';
import { ManufacturerDto } from './ManufacturerDto';

export abstract class CustomerDto {
  $type: string;
  id: number;
  managerId: number;
  customerType: CustomerType;
  customerCategoryId: number;
  name: string;
  imageUrl: string;
  email: string;
  phone: string;
  fax: string;
  note: string;

  invoiceTo: string;
  invoiceAddress: string;
  invoiceSuburb: string;
  invoiceState: string;
  invoicePostcode: string;

  deliveryTo: string;
  deliveryAddress: string;
  deliverySuburb: string;
  deliveryState: string;
  deliveryPostcode: string;

  discountRate: number;
  deliveryFee: number;

  isOnHold: boolean;

  @Type(() => CustomerCategoryDto, {
    keepDiscriminatorProperty: true,
    discriminator: {
      property: '$type',
      subTypes: [
        {
          value: CustomerCategoryAccountDto,
          name: '_3_Application.Dtos.CustomerCategory.CustomerCategoryAccountDto, 3-Application',
        },
        {
          value: CustomerCategoryCBDDto,
          name: '_3_Application.Dtos.CustomerCategory.CustomerCategoryCBDDto, 3-Application',
        },
      ],
    },
  })
  customerCategory: CustomerCategoryDto;

  get invoiceDetails(): string {
    return `${this.invoiceAddress}, ${this.invoiceSuburb} ${this.invoiceState} ${this.invoicePostcode}`;
  }

  get deliveryDetails(): string {
    return `${this.deliveryAddress}, ${this.deliverySuburb} ${this.deliveryState} ${this.deliveryPostcode}`;
  }

  constructor() {}
}

File: JanperCab/3-Application/Dtos/Customer/CustomerDto.cs
using _1_Domain.Enum;
using _3_Application.Dtos.CustomerCategory;
using System.ComponentModel.DataAnnotations;

namespace _3_Application.Dtos.Customer
{
    public abstract class CustomerDto
    {
        public int Id { get; set; }

        public int? ManagerId { get; set; }

        public CustomerType CustomerType { get; set; }

        public int CustomerCategoryId { get; set; }

        [Required]
        [StringLength(255)]
        public string Name { get; set; }

        public string ImageUrl { get; set; }

        [Required]
        [StringLength(255)]
        public string Email { get; set; }

        [Required]
        [StringLength(255)]
        public string Phone { get; set; }

        [StringLength(255)]
        public string Fax { get; set; }

        public string Note { get; set; }


        public string InvoiceTo { get; set; }

        public string InvoiceAddress { get; set; }

        public string InvoiceSuburb { get; set; }

        public string InvoiceState { get; set; }

        public string InvoicePostcode { get; set; }

        public string DeliveryTo { get; set; }

        public string DeliveryAddress { get; set; }

        public string DeliverySuburb { get; set; }

        public string DeliveryState { get; set; }

        public string DeliveryPostcode { get; set; }



        public decimal DiscountRate { get; set; }

        public decimal DeliveryFee { get; set; }

        public bool IsOnHold { get; set; }


        public CustomerCategoryDto CustomerCategory { get; set; }

        public CustomerDto Manager { get; set; }
    }
}